drivers/sensor/amd_sb_tsi/sb_tsi_emul.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT amd_sb_tsi
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/emul.h>
+#include <zephyr/drivers/emul_sensor.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c_emul.h>
+#include <zephyr/logging/log.h>
+#include "sb_tsi.h"
+
+LOG_MODULE_DECLARE(AMD_SB_TSI, CONFIG_SENSOR_LOG_LEVEL);
+
+#define NUM_REGS 128
+
+struct sb_tsi_emul_data {
+	uint8_t reg[NUM_REGS];
+};
+
+struct sb_tsi_emul_cfg {
+};
+
+static void sb_tsi_emul_set_reg(const struct emul *target, uint8_t reg, uint8_t val)
+{
+	struct sb_tsi_emul_data *data = target->data;
+
+	__ASSERT_NO_MSG(reg < NUM_REGS);
+	data->reg[reg] = val;
+}
+
+static uint8_t sb_tsi_emul_get_reg(const struct emul *target, uint8_t reg)
+{
+	struct sb_tsi_emul_data *data = target->data;
+
+	__ASSERT_NO_MSG(reg < NUM_REGS);
+	return data->reg[reg];
+}
+
+static void sb_tsi_emul_reset(const struct emul *target)
+{
+	struct sb_tsi_emul_data *data = target->data;
+
+	memset(data->reg, 0, NUM_REGS);
+}
+
+static int sb_tsi_emul_transfer_i2c(const struct emul *target, struct i2c_msg *msgs,
+				    int num_msgs, int addr)
+{
+	/* Largely copied from emul_bmi160.c */
+	unsigned int val;
+	int reg;
+
+	__ASSERT_NO_MSG(msgs && num_msgs);
+
+	i2c_dump_msgs_rw("emul", msgs, num_msgs, addr, false);
+	switch (num_msgs) {
+	case 2:
+		if (msgs->flags & I2C_MSG_READ) {
+			LOG_ERR("Unexpected read");
+			return -EIO;
+		}
+		if (msgs->len != 1) {
+			LOG_ERR("Unexpected msg0 length %d", msgs->len);
+			return -EIO;
+		}
+		reg = msgs->buf[0];
+
+		/* Now process the 'read' part of the message */
+		msgs++;
+		if (msgs->flags & I2C_MSG_READ) {
+			switch (msgs->len) {
+			case 1:
+				val = sb_tsi_emul_get_reg(target, reg);
+				msgs->buf[0] = val;
+				break;
+			default:
+				LOG_ERR("Unexpected msg1 length %d", msgs->len);
+				return -EIO;
+			}
+		} else {
+			if (msgs->len != 1) {
+				LOG_ERR("Unexpected msg1 length %d", msgs->len);
+			}
+			sb_tsi_emul_set_reg(target, reg, msgs->buf[0]);
+		}
+		break;
+	default:
+		LOG_ERR("Invalid number of messages: %d", num_msgs);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int sb_tsi_emul_init(const struct emul *target, const struct device *parent)
+{
+	sb_tsi_emul_reset(target);
+	return 0;
+}
+
+static int sb_tsi_emul_set_channel(const struct emul *target, enum sensor_channel chan,
+				   q31_t value, int8_t shift)
+{
+	struct sb_tsi_emul_data *data = target->data;
+	int64_t scaled_value;
+	int temp_eighths;
+
+	if (chan != SENSOR_CHAN_DIE_TEMP) {
+		return -ENOTSUP;
+	}
+
+	scaled_value = (int64_t)value << shift;
+	temp_eighths = scaled_value >> 28;
+	temp_eighths = CLAMP(temp_eighths, 0, 0x7ff);
+
+	data->reg[SB_TSI_TEMP_INT] = temp_eighths >> 3;
+	data->reg[SB_TSI_TEMP_DEC] = (temp_eighths & 0x7) << 5;
```

Comment:
Continuing on from the other comment, what you want to do here is convert the fixed-point Celsius value passed in by the test to its corresponding register representation. In this case I'd unshift the supplied value and apply the appropriate scaling factor to convert to integer millicelsius. Then divide by `1/0.125 = 8` and clamp to get the register value.

Reply:
Is this better?